### PR TITLE
feat(admin): allow groups for hierarchical inclusion

### PR DESCRIPTION
**Pull request by HejdaJakub**
* Perun admin can dis/allow groups of some member organization for hierarchical inclusion to specific parent vo.
* Hierarchical inclusion has its own card (in advanced settings) and this card is only visible if the organization is member of at least one parent organization.

### CODE REVIEW

File: apps/admin-gui/src/app/shared/side-menu/side-menu-item.service.ts [hunk 1]
```diff
@@ -576,6 +576,15 @@ export class SideMenuItemService {
         });
       }
 
+      // Hierarchical inclusion
+      if (this.authResolver.isPerunAdmin()) {
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/side-menu/side-menu-item.service.ts:580]: Also check if this organization have some parent.

File: apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import { SelectionModel } from '@angular/cdk/collections';
+import { getDefaultDialogConfig } from '@perun-web-apps/perun/utils';
+import { MatDialog } from '@angular/material/dialog';
+import { AddGroupHierarchicalIncludeDialogComponent } from '../../../../../shared/components/dialogs/add-group-hierarchical-include-dialog/add-group-hierarchical-include-dialog.component';
+import { UniversalConfirmationItemsDialogComponent } from '@perun-web-apps/perun/dialogs';
+import { TranslateService } from '@ngx-translate/core';
+import { TABLE_HIERARCHICAL_INCLUSION } from '@perun-web-apps/config/table-config';
+
+@Component({
+  selector: 'app-perun-web-apps-vo-settings-hierarchical-inclusion',
```
Comment (bodnara) [inline apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.ts:20]:
```suggestion
  selector: 'app-vo-settings-hierarchical-inclusion',
```

File: apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  loading = false;
+  vo: Vo;
+  parentVos: Vo[] = [];
+  selectedParentVo: Vo;
+  allowedGroups: Group[] = [];
+  selected = new SelectionModel<RichGroup>(true, []);
+  tableId = TABLE_HIERARCHICAL_INCLUSION;
+
+  constructor(
+    private route: ActivatedRoute,
```
Comment (bodnara) [inline apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.ts:34]:
```suggestion

```

File: apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  templateUrl: './vo-settings-hierarchical-inclusion.component.html',
+  styleUrls: ['./vo-settings-hierarchical-inclusion.component.scss'],
+})
+export class VoSettingsHierarchicalInclusionComponent implements OnInit {
+  loading = false;
+  vo: Vo;
+  parentVos: Vo[] = [];
+  selectedParentVo: Vo;
+  allowedGroups: Group[] = [];
+  selected = new SelectionModel<RichGroup>(true, []);
```
Comment (bodnara) [inline apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.ts:30]:
```suggestion
  selected = new SelectionModel<Group>(true, []);
```

File: apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.html [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  </div>
+</div>
+<perun-web-apps-groups-list
+  *ngIf="!loading"
+  [tableId]="tableId"
+  [groups]="allowedGroups"
+  [selection]="selected"
+  [displayedColumns]="['select', 'id', 'name', 'description']"
+  [noGroupsAlert]="'VO_DETAIL.SETTINGS.HIERARCHICAL_INCLUSION.NO_GROUPS_ALLOWED_ALERT'"
+  (refreshTable)="loadAllowedGroups()"
```
Comment (bodnara) [inline apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.html:30]: As of right now there is no way to trigger this refresh based on the displayed columns.

File: apps/admin-gui/src/app/shared/components/dialogs/add-group-hierarchical-include-dialog/add-group-hierarchical-include-dialog.component.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+export interface AddGroupHierarchicalIncludeDialogData {
+  theme: string;
+  voId: number;
+  parentVo: Vo;
+  allowedGroupsIds: number[];
+}
+
+@Component({
+  selector: 'app-perun-web-apps-add-group-hierarchical-include-dialog',
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/add-group-hierarchical-include-dialog/add-group-hierarchical-include-dialog.component.ts:17]:
```suggestion
  selector: 'app-add-group-hierarchical-include-dialog',
```

File: apps/admin-gui/src/app/shared/components/dialogs/add-group-hierarchical-include-dialog/add-group-hierarchical-include-dialog.component.html [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    <mat-spinner *ngIf="loading" class="ml-auto mr-auto"></mat-spinner>
+  </div>
+  <div *ngIf="!loading" mat-dialog-actions>
+    <button (click)="close()" mat-flat-button class="ml-auto">
+      {{'DIALOGS.ADD_GROUPS_HIERARCHICAL_INCLUSION.CANCEL' | translate}}
+    </button>
+    <button
+      [disabled]="selected.selected.length === 0"
+      (click)="confirm()"
+      class="ml-2"
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/add-group-hierarchical-include-dialog/add-group-hierarchical-include-dialog.component.html:35]: This margin is not needed, instead some margin could be set on the cancel button.

File: apps/admin-gui/src/app/shared/components/dialogs/add-group-hierarchical-include-dialog/add-group-hierarchical-include-dialog.component.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  templateUrl: './add-group-hierarchical-include-dialog.component.html',
+  styleUrls: ['./add-group-hierarchical-include-dialog.component.scss'],
+})
+export class AddGroupHierarchicalIncludeDialogComponent implements OnInit {
+  loading = false;
+  theme: string;
+  voId: number;
+  parentVo: Vo;
+  groups: Group[];
+  selected = new SelectionModel<RichGroup>(true, []);
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/add-group-hierarchical-include-dialog/add-group-hierarchical-include-dialog.component.ts:27]:
```suggestion
  selected = new SelectionModel<Group>(true, []);
```

File: apps/admin-gui/src/app/shared/components/dialogs/add-group-hierarchical-include-dialog/add-group-hierarchical-include-dialog.component.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  tableId = TABLE_ADD_HIERARCHICAL_INCLUSION;
+  filterValue = '';
+
+  constructor(
+    private dialogRef: MatDialogRef<AddGroupHierarchicalIncludeDialogComponent>,
+    @Inject(MAT_DIALOG_DATA) public data: AddGroupHierarchicalIncludeDialogData,
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/add-group-hierarchical-include-dialog/add-group-hierarchical-include-dialog.component.ts:33]:
```suggestion
    @Inject(MAT_DIALOG_DATA) private data: AddGroupHierarchicalIncludeDialogData,
```

**Review by bodnara — CHANGES_REQUESTED** [review]
I could be helpful if the organization name was highlighted in some form (e.g. bold) in both the add and remove dialog.

File: apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.html (hunk 5 above)
Comment (HejdaJakub) [reply apps/admin-gui/src/app/vos/pages/vo-detail-page/vo-settings/vo-settings-hierarchical-inclusion/vo-settings-hierarchical-inclusion.component.html]: `refreshTable` attribute removed from html

**Review by bodnara — APPROVED** [review]

**Review by xkureck — CHANGES_REQUESTED** [review]
image
The "Parent organization Radek's Tenant will NOT be able to include these groups." text loks kind of weird to me, but thats maybe of that bug on app-alert I guess?
Also on the red button I would put the label "Remove" than "Submit"

File: apps/admin-gui/src/assets/i18n/en.json [hunk 10]
```diff
@@ -1131,6 +1138,23 @@
       "SELECT_VO": "Select organization",
       "FILTER_GROUPS": "Filter groups by Id, name or description"
     },
+    "ADD_GROUPS_HIERARCHICAL_INCLUSION": {
+      "TITLE": "Allow groups for hierarchical inclusion",
+      "INFO_START": "Selected groups will be allowed for hierarchical inclusion. Parent organization ",
+      "INFO_END": " will be able to include these groups (including group members).",
```
[the remark this answers is not in the document]
Comment (HejdaJakub) [reply apps/admin-gui/src/assets/i18n/en.json]: Cool, I didn't know about this translation behavior with params. Thanks!

**Comment by HejdaJakub** [comment]
The "Parent organization Radek's Tenant will NOT be able to include these groups." text loks kind of weird to me, but thats maybe of that bug on app-alert I guess? Also on the red button I would put the label "Remove" than "Submit"

You're right, that's due to the app-alert bug. I rebased it, so now it looks correct.

File: apps/admin-gui/src/assets/i18n/en.json [hunk 11]
```diff
@@ -1131,6 +1138,23 @@
...
+      "NO_GROUPS_FOUND_ALERT": "No groups found.",
+      "CANCEL": "Cancel",
+      "ALLOW": "Allow",
+      "SUCCESS": "Groups were successfully allowed"
+    },
+    "REMOVE_GROUPS_HIERARCHICAL_INCLUSION": {
+      "TITLE": "Disallow groups for hierarchical inclusion",
+      "DESCRIPTION": "Selected groups will be disallowed for hierarchical inclusion.",
+      "ALERT_START": "Parent organization ",
+      "ALERT_END": " will NOT be able to include these groups.",
```
Comment (xkureck) [inline apps/admin-gui/src/assets/i18n/en.json:1154]:
```suggestion

```

**Review by xkureck — CHANGES_REQUESTED** [review]
Just one small thing, then we can merge it

**Review by xkureck — APPROVED** [review]

**Pull request merged**